pkg/service/backup/restore_worker_data.go

```diff
@@ -0,0 +1,544 @@
+func (ctr SSTableCounter) RestoreTables(v int) bool {
+	return ctr.Sub(int64(v)) == 0
+}
+
+// counter creates and initializes SSTableCounter
+// for currently restored table.
+func (w *restoreWorker) newCounter() SSTableCounter {
+	ctr := NewSSTableCounter(len(w.bundleIDPool))
+
+	for _, h := range w.hosts {
+		if h.OngoingJob != nil {
+			ctr.AddTables(len(h.OngoingJob.SSTableID))
+		}
+	}
+
+	return ctr
+}
+
+// workFunc is responsible for creating and restoring batches on multiple hosts (possibly in parallel).
+// It requires previous initialization of restore worker components.
+func (w *restoreWorker) workFunc(ctx context.Context, run *RestoreRun, target RestoreTarget, fm FilesMeta) error {
+	version, err := w.RecordTableVersion(ctx, fm.Keyspace, fm.Table)
+	if err != nil {
+		return err
+	}
+
+	var (
+		srcDir = w.location.RemotePath(w.miwc.SSTableVersionDir(fm.Keyspace, fm.Table, fm.Version))
+		dstDir = uploadTableDir(fm.Keyspace, fm.Table, version)
+	)
+
+	w.Logger.Info(ctx, "Found source and destination directory",
+		"src_dir", srcDir,
+		"dst_dir", dstDir,
+	)
+
+	ctr := w.newCounter()
+
+	// Every host has its personal goroutine which is responsible
+	// for creating and downloading batches.
+	// Goroutine returns only in case of error or
+	// if the whole table has been restored.
+	return parallel.Run(len(w.hosts), target.Parallel, func(n int) error {
+		for {
+			if ctx.Err() != nil {
+				return parallel.Abort(ctx.Err())
+			}
+			// Current goroutine's host
+			h := &w.hosts[n]
+
+			pr, err := w.createJob(ctx, run, target, h, srcDir, dstDir)
+			if err != nil {
+				return errors.Wrapf(err, "create run progress for host: %s", h.Host)
+			}
+			if pr == nil {
+				w.Logger.Info(ctx, "No more batches to restore", "host", h.Host)
+				return nil
+			}
+
+			batch := w.batchFromIDs(pr.SSTableID)
+
+			w.Logger.Info(ctx, "Waiting for job",
+				"host", h.Host,
+				"job_id", pr.AgentJobID,
+			)
+
+			if err := w.waitJob(ctx, pr); err != nil {
+				// In case of context cancellation restore is interrupted
+				if ctx.Err() != nil {
+					return parallel.Abort(ctx.Err())
+				}
+				// As run progress could have already been inserted
+				// into the database, it should be deleted.
+				w.deleteRunProgress(ctx, pr)
+				w.returnBatchToPool(pr.SSTableID)
+
+				return errors.Wrapf(err, "wait on rclone job, id: %d, host: %s", pr.AgentJobID, h.Host)
+			}
+
+			w.Logger.Info(ctx, "Calling agent's restore", "host", h.Host)
+
+			if err := w.Client.Restore(ctx, h.Host, fm.Keyspace, fm.Table, version, batch); err != nil {
+				w.deleteRunProgress(ctx, pr)
+				w.returnBatchToPool(pr.SSTableID)
+
+				return errors.Wrap(err, "call agent's restore")
+			}
+
+			w.Logger.Info(ctx, "Restored batch",
+				"host", h.Host,
+				"batch", batch,
+			)
+
+			// Close pool and free hosts awaiting on it if all SSTables have been successfully restored.
+			if ctr.RestoreTables(len(pr.SSTableID)) {
```

Comment:
nit: `RestoreTables` is not descriptive enough.
It decreases the counter and checks if counter == 0.
Maybe the previous approach without this method, but the check maybe explicitly here was cleaner.